### PR TITLE
OCS-580: Respin ceph pod(osd, mon, mgr) and check the interaction with monitoring pods

### PR DESCRIPTION
**Pull request by Akarsha-rai**
A testcase to validate respinning ceph pods(osd, mon, mgr) shouldn't have functional impact to prometheus pods, all data/metrics should be collected correctly.

Signed-off-by: Akarsha <akrai@redhat.com>

### CODE REVIEW

**Comment by Akarsha-rai** [comment]
[test_log.txt](https://github.com/red-hat-storage/ocs-ci/files/3468266/test_log.txt)


============================================================================ 
                      short test summary info ============================================================================
PASSED tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py::TestRespinCephPodsAndInteractionWithPrometheus::test_respinning_ceph_pods_and_interaction_with_prometheus_pod
============================================================================1 passed in 676.70 seconds ============================================================================

File: ocs_ci/ocs/monitoring.py [hunk 1]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
+
+    Args:
+        pvc_name (str): Name of the pvc
+
+    """
+    logger.info(
+        f"Verify for created pvc {pvc_name} related data is collected on prometheus pod"
+    )
+    pvcs_data = get_kube_pod_spec_volumes_persistentvolumeclaims_info_metric()
+    list_pvcs_data = pvcs_data['data']['result']
```
Comment (ebenahar) [inline ocs_ci/ocs/monitoring.py]:
```python
pvcs_data.get('data').get('result')
```

File: ocs_ci/ocs/monitoring.py [hunk 2]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
+    pvc_list = [pvc for pvc in list_pvcs_data if pvc_name == pvc['metric']['persistentvolumeclaim']]
```
Comment (ebenahar) [inline ocs_ci/ocs/monitoring.py]:
```python
pvc.get('metric').get('persistentvolumeclaim')
```

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+
+def create_multiple_project_and_pvc_and_check_metrics_are_collected(self):
+    """
+    Creates projects, pvcs and app pods
+    """
+    for i in range(5):
+        # Create new project
+        self.namespace = create_unique_resource_name('test', 'namespace')
+        self.project_obj = ocp.OCP(kind='Project', namespace=self.namespace)
```
Comment (ebenahar) [inline tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: please use helpers.create_project()

**Comment by Akarsha-rai** [comment]
[test_log.txt](https://github.com/red-hat-storage/ocs-ci/files/3477343/test_log.txt)

============================================================================ 
  short test summary info ============================================================================
PASSED tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py::TestRespinCephPodsAndInteractionWithPrometheus::test_respinning_ceph_pods_and_interaction_with_prometheus_pod
============================================================================1 passed in 685.32 seconds ============================================================================

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py (hunk 3 above)
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: done

File: ocs_ci/ocs/monitoring.py (hunk 2 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: done

File: ocs_ci/ocs/monitoring.py (hunk 1 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: done

File: ocs_ci/ocs/monitoring.py [hunk 4]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
         pvc_dict = ocp_pvc_obj.get(resource_name=pvc.get('metadata').get('name'))
         pvc_obj = PVC(**pvc_dict)
         pvc_obj_list.append(pvc_obj)
     return pvc_obj_list
+
+
+def get_kube_pod_spec_volumes_persistentvolumeclaims_info_metric():
+    """
+    Returns the created pvc information
+
```
Comment (RazTamir) [inline ocs_ci/ocs/monitoring.py]: add return value

File: ocs_ci/ocs/monitoring.py [hunk 5]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
+
+@retry(UnexpectedBehaviour, tries=10, delay=3, backoff=1)
+def collected_metrics_for_created_pvc(pvc_name):
+    """
+    Checks whether initially pvc related data is collected on pod
+
+    Args:
+        pvc_name (str): Name of the pvc
+
```
Comment (RazTamir) [inline ocs_ci/ocs/monitoring.py]: add return/raise values

File: ocs_ci/ocs/monitoring.py [hunk 6]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
+    """
+    logger.info(
+        f"Verify for created pvc {pvc_name} related data is collected on prometheus pod"
+    )
+    pvcs_data = get_kube_pod_spec_volumes_persistentvolumeclaims_info_metric()
+    list_pvcs_data = pvcs_data.get('data').get('result')
+    pvc_list = [pvc for pvc in list_pvcs_data if pvc_name == pvc.get('metric').get('persistentvolumeclaim')]
+    if not pvc_list:
+        raise UnexpectedBehaviour(
```
Comment (RazTamir) [inline ocs_ci/ocs/monitoring.py]: I would log the error and return False if the intention is that this class will return bool

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    # Switch to default project
+    ret = ocp.switch_to_default_rook_cluster_project()
+    assert ret, 'Failed to switch to default rook cluster project'
+
+    # Delete projects created
+    for prj in self.namespace_list:
+        self.prj_obj.delete(resource_name=prj)
+
+
+def create_multiple_project_and_pvc_and_check_metrics_are_collected(self):
```
Comment (RazTamir) [inline tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: create_multiple_projects AND create_pvc AND check_metrics_are_collected.
you have 3 functions here - please split

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    self.pvc_objs = []
+    self.pod_objs = []
+
+    assert create_multiple_project_and_pvc_and_check_metrics_are_collected(self)
+
+
+def teardown(self):
+    """
+    Delete app pods and PVCs
+    Delete project
```
Comment (vavuthu) [inline tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: Delete app pods, PVCs and project

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+    # Delete projects created
+    for prj in self.namespace_list:
+        self.prj_obj.delete(resource_name=prj)
+
+
+def create_multiple_project_and_pvc_and_check_metrics_are_collected(self):
+    """
+    Creates projects, pvcs and app pods
+    """
+    for i in range(5):
```
Comment (vavuthu) [inline tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: If you are not using i, use _

File: ocs_ci/ocs/monitoring.py (hunk 6 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: Here I'm using retry decorator so raising exceptions.

File: ocs_ci/ocs/monitoring.py (hunk 5 above)
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: done

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py (hunk 7 above)
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: done

File: ocs_ci/ocs/monitoring.py [hunk 10]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
+    """
+    prometheus = ocs_ci.utility.prometheus.PrometheusAPI()
+    response = prometheus.get(
+        'query?query=kube_pod_spec_volumes_persistentvolumeclaims_info'
+    )
+    return json.loads(response.content.decode('utf-8'))
+
+
+@retry(UnexpectedBehaviour, tries=10, delay=3, backoff=1)
+def collected_metrics_for_created_pvc(pvc_name):
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: renamed

File: ocs_ci/ocs/monitoring.py [hunk 11]
```diff
@@ -53,49 +58,66 @@ def validate_pvc_created_and_bound_on_monitoring_pods():
...
     for pvc in pvc_list['items']:
         pvc_dict = ocp_pvc_obj.get(resource_name=pvc.get('metadata').get('name'))
         pvc_obj = PVC(**pvc_dict)
         pvc_obj_list.append(pvc_obj)
     return pvc_obj_list
+
+
+def get_kube_pod_spec_volumes_persistentvolumeclaims_info_metric():
+    """
+    Returns the created pvc information
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply ocs_ci/ocs/monitoring.py]: done

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture()
+def test_fixture(request):
+    """
+    Setup and teardown
+    """
+    self = request.node.cls
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: removed fixtures using class references

File: tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
+import logging
+
```
[the remark this answers is not in the document]
Comment (Akarsha-rai) [reply tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py]: done

File: ocs_ci/ocs/monitoring.py (hunk 5 above)
Comment (RazTamir) [reply ocs_ci/ocs/monitoring.py]: also raise UnexpectedBehaviour

File: ocs_ci/ocs/monitoring.py (hunk 6 above)
Comment (RazTamir) [reply ocs_ci/ocs/monitoring.py]: ack

**Comment by Akarsha-rai** [comment]
[test_log.txt](https://github.com/red-hat-storage/ocs-ci/files/3524755/test_log.txt)

=============================================================================short test summary info =============================================================================
PASSED tests/e2e/monitoring/test_when_respin_ceph_pods_interaction_with_prometheus.py::TestRespinCephPodsAndInteractionWithPrometheus::test_monitoring_after_respinning_ceph_pods
=============================================================================1 passed in 437.21 seconds =============================================================================

File: ocs_ci/ocs/monitoring.py (hunk 5 above)
Comment (RazTamir) [reply ocs_ci/ocs/monitoring.py]: @Akarsha-rai could you please add the raise UnexpectedBehaviour ?

**Comment by keesturam** [comment]
LGTM.

**Pull request merged**